File: src-tauri/src/utils/websocket.rs
// src-tauri/src/websocket.rs

use crate::utils::audio::read_audio_file;
use futures_util::{stream::SplitSink, SinkExt, StreamExt};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::net::{TcpListener, TcpStream};
use tokio::sync::Mutex;
use tokio_tungstenite::{accept_async, tungstenite::Message};
type WebSocket = tokio_tungstenite::WebSocketStream<TcpStream>;
type WebSocketWriter = SplitSink<WebSocket, Message>;
type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;
use serde::{Deserialize, Serialize};

#[derive(Default)]
struct ConnectionManager {
    connections: Mutex<HashMap<u32, WebSocketWriter>>,
}

#[derive(Deserialize, Serialize)]
struct ReadAudioFileRequest {
    command: String,
    path: String,
}

impl ConnectionManager {
    async fn add_connection(&self, id: u32, writer: WebSocketWriter) {
        self.connections.lock().await.insert(id, writer);
    }

    async fn remove_connection(&self, id: &u32) {
        self.connections.lock().await.remove(id);
    }

    async fn send_message(&self, id: &u32, msg: Message) -> Result<()> {
        if let Some(writer) = self.connections.lock().await.get_mut(id) {
            writer.send(msg).await?;
        }
        Ok(())
    }
}

pub async fn start_server() -> Result<()> {
    let addr = "127.0.0.1:8080".to_string();
    let listener = TcpListener::bind(&addr).await?;

    let manager = Arc::new(ConnectionManager::default());

    while let Ok((stream, _)) = listener.accept().await {
        let manager = manager.clone();
        tokio::spawn(async move {
            if let Err(e) = handle_connection(manager, stream).await {
                eprintln!("Error handling connection: {}", e);
            }
        });
    }

    Ok(())
}

async fn handle_connection(manager: Arc<ConnectionManager>, stream: TcpStream) -> Result<()> {
    let ws_stream = accept_async(stream).await?;
    let (writer, mut reader) = ws_stream.split();
    let id = rand::random::<u32>();

    manager.add_connection(id, writer).await;

    while let Some(Ok(msg)) = reader.next().await {
        match msg {
            Message::Text(text) => {
                println!("Received a text message: {}", text);
                if let Ok(request) = serde_json::from_str::<ReadAudioFileRequest>(&text) {
                    if request.command == "read_audio_file" {
                        match read_audio_file(request.path) {
                            Ok(data) => {
                                // 分块传输文件数据
                                let chunk_size = 14 * 1024 * 1024; // 8 MB per chunk
                                for chunk in data.chunks(chunk_size) {
                                    let response = Message::Binary(chunk.to_vec());
                                    manager.send_message(&id, response).await?;
                                }
                            }
                            Err(e) => {
                                let error_msg = format!("Error reading audio file: {}", e);
                                manager.send_message(&id, Message::Text(error_msg)).await?;
                            }
                        }
                    }
                }
            }

            Message::Binary(bin) => {
                println!("Received a binary message: {:?}", bin);
            }
            Message::Close(_) => {
                manager.remove_connection(&id).await;
                break;
            }
            _ => {}
        }
    }

    Ok(())
}

File: src/app/services/websocket.ts
import WebSocket from '@tauri-apps/plugin-websocket';

export async function connectWebSocket() {
    try {
        const ws = await WebSocket.connect('ws://127.0.0.1:8080');
        let fileData: any = [];

        ws.addListener((msg) => {
            if (msg.type === 'Text') {
                console.log('Received Text Message:', msg.data);
            } else if (msg.type === 'Binary') {
                console.log('Received Binary Message:', new Uint8Array(msg.data));
                fileData.push(new Uint8Array(msg.data));
            } else {
                console.log('Received Message:', msg);
            }
        });

        const request = {
            command: "read_audio_file",
            path: ""
        };

        await ws.send(JSON.stringify(request));

        // 添加适当的延迟，以确保消息能被接收
        setTimeout(async () => {
            await ws.disconnect();

            // 处理接收到的完整文件数据
            const totalLength = fileData.reduce((acc, curr) => acc + curr.length, 0);
            const combinedArray = new Uint8Array(totalLength);
            let offset = 0;
            for (const chunk of fileData) {
                combinedArray.set(chunk, offset);
                offset += chunk.length;
            }

            console.log('Complete file data:', combinedArray);
        }, 10000); // 10秒后断开连接

    } catch (error) {
        console.error('WebSocket connection error:', error);
    }
}

export async function fetchAudioFile(path: string) {
    try {
        const ws = await WebSocket.connect('ws://127.0.0.1:8080');
        let fileData = [];

        ws.addListener((msg) => {
            if (msg.type === 'Text') {
                console.log('Received Text Message:', msg.data);
            } else if (msg.type === 'Binary') {
                console.log('Received Binary Message:', new Uint8Array(msg.data));
                fileData.push(new Uint8Array(msg.data));
            } else {
                console.log('Received Message:', msg);
            }
        });

        const request = {
            command: "read_audio_file",
            path: path
        };

        await ws.send(JSON.stringify(request));

        // 添加适当的延迟，以确保消息能被接收
        setTimeout(async () => {
            await ws.disconnect();

            // 处理接收到的完整文件数据
            const totalLength = fileData.reduce((acc, curr) => acc + curr.length, 0);
            const combinedArray = new Uint8Array(totalLength);
            let offset = 0;
            for (const chunk of fileData) {
                combinedArray.set(chunk, offset);
                offset += chunk.length;
            }

            console.log('Complete file data:', combinedArray);
            return combinedArray;
        }, 10000); // 10秒后断开连接

    } catch (error) {
        console.error('WebSocket connection error:', error);
    }
}